pkg/tfbridge/provider.go

```diff
@@ -375,43 +536,17 @@ func (p *Provider) Configure(ctx context.Context,
 		}
 	}
 
-	// So we can provide better error messages, do a quick scan of required configs for this
-	// schema and report any that haven't been supplied.
-	var missingKeys []*pulumirpc.ConfigureErrorMissingKeys_MissingKey
-	for key, meta := range p.config {
-		if meta.Required && !config.IsSet(key) {
-			name := TerraformToPulumiName(key, meta, nil, false)
-			fullyQualifiedName := tokens.NewModuleToken(p.pkg(), tokens.ModuleName(name))
-
-			// TF descriptions often have newlines in inopportune positions. This makes them present
-			// a little better in our console output.
-			descriptionWithoutNewlines := strings.Replace(meta.Description, "\n", " ", -1)
-			missingKeys = append(missingKeys, &pulumirpc.ConfigureErrorMissingKeys_MissingKey{
-				Name:        fullyQualifiedName.String(),
-				Description: descriptionWithoutNewlines,
-			})
-		}
-	}
-
-	if len(missingKeys) > 0 {
-		// Clients of our RPC endpoint will be looking for this detail in order to figure out
-		// which keys need descriptive error messages.
-		err = rpcerror.WithDetails(
-			rpcerror.New(codes.InvalidArgument, "required configuration keys were missing"),
-			&pulumirpc.ConfigureErrorMissingKeys{MissingKeys: missingKeys})
-		return nil, err
-	}
-
-	// Perform validation of the config state so we can offer nice errors.
-	warns, errs := p.tf.Validate(config)
-	for _, warn := range warns {
-		if err = p.host.Log(ctx, diag.Warning, "", fmt.Sprintf("provider config warning: %v", warn)); err != nil {
+	if req.Variables == nil {
```

Comment:
Can we factor out any of the code above? It looks like at least some of it can be shared between `CheckConfig`/`Configure`, and I think the `PreConfigureCallback` should only be called once.

Reply:
Ok, I've refactored as much as I can here. I also moved the `PreConfigureCallback` to be inside the `if req.Varariables == nil` in the `Configure` func to ensure that we only call it if we are hitting the old (non CheckConfig) path via the CLI